src/ast/passes/map_sugar.cpp

```diff
@@ -0,0 +1,274 @@
+public:
+  using Visitor<MapAssignmentCheck>::visit;
+  void visit(Call &call);
+};
+
+} // namespace
+
+// These are special functions which are part of the map API, and operate
+// independently of any key. We allow the first argument to be a pure map,
+// and process therefore don't expand a default key in these cases.
+//
+// In the future, this could be generalized by extracting information about the
+// specific function being called, and potentially respecting annotations on
+// these arguments.
+static std::unordered_set<std::string> RAW_MAP_ARG = {
+  "print", "clear", "zero", "len", "delete", "has_key",
+};
+
+void MapDefaultKey::visit(Map &map)
+{
+  check(map, false);
+}
+
+void MapDefaultKey::visit(MapAccess &acc)
+{
+  check(*acc.map, true);
+  visitAndReplace(&acc.key);
+}
+
+void MapDefaultKey::visit(AssignScalarMapStatement &assign)
+{
+  check(*assign.map, false);
+  visitAndReplace(&assign.expr);
+}
+
+void MapDefaultKey::visit(AssignMapStatement &assign)
+{
+  check(*assign.map, true);
+  visitAndReplace(&assign.key);
+  visitAndReplace(&assign.expr);
+}
+
+void MapDefaultKey::check(Map &map, bool indexed)
+{
+  bool scalar = !indexed;
+  auto val = metadata_.scalar.find(map.ident);
+  if (val == metadata_.scalar.end()) {
+    metadata_.scalar.emplace(map.ident, scalar);
+  } else if (val->second != scalar) {
+    if (!scalar) {
+      map.addError() << map.ident
+                     << " used as a non-scalar map, previously used as a "
+                        "scalar map";
+    } else {
+      map.addError()
+          << map.ident
+          << " used as a scalar map, previously used as a non-scalar map";
+    }
+  }
+}
+
+void MapDefaultKey::visit(Call &call)
+{
+  // Skip the first argument in these cases. This allows the argument to be
+  // *either* a pure map, or a map access. Later passes will figure out what to
+  // do with this, as they may have parametric behavior (as with print).
+  if (RAW_MAP_ARG.contains(call.func) && !call.vargs.empty()) {
+    if (auto *map = dynamic_cast<Map *>(call.vargs.at(0))) {
+      // Check our functions for consistency. These are effectively builtins
+      // that require that map to have keys (conditionally for delete).
+      if (call.func == "delete") {
+        check(*map, call.vargs.size() > 1);
+      } else if (call.func == "has_key") {
+        check(*map, true);
+      } else if (call.func == "len") {
+        check(*map, true);
+      }
+    } else {
+      // Process the argument.
+      visitAndReplace(&call.vargs.at(0));
+    }
+    for (size_t i = 1; i < call.vargs.size(); i++) {
+      visitAndReplace(&call.vargs.at(i));
+    }
+  } else {
+    Visitor<MapDefaultKey>::visit(call);
+  }
+}
+
+void MapDefaultKey::visit(For &for_loop)
+{
+  check(*for_loop.map, true);
+  Visitor<MapDefaultKey>::visit(for_loop.stmts);
+}
+
+void MapFunctionAliases::visit(Call &call)
+{
+  // If this is specfically `delete(@)` then we remap this to `clear`. We
```

Comment:
The problem with that is you've turned a sync function into an async function. Can't we just desugar it to `delete(@, key)` or for scalars `delete(@, 0)` ?

Reply:
Yikes, I didn't even realize that `clear` was async. Yes, I can desugar as `delete(..., 0)`.

Reply:
So I originally got this from the docs, which I guess are wrong:

```
Delete a single key from a map. For scalar maps (e.g. no explicit keys), the key is omitted and is equivalent to calling clear.
```

It is *not* equivalent to `clear`.

Reply:
> It is not equivalent to clear.

It certainly is not.

Reply:
This transformation is no longer being done.